### ISSUE TITLE
Not all codec configurations can currently be configured with PinConfigurations

### ISSUE THREAD

**Issue by vit9696** [acidanthera/bugtracker#730]
For some reason on some configurations it is not enough to copy-paste commands to HDA controller from software like CodecCommander to AppleALC PinConfigurations to get audio work properly. The reason for this is unknown and requires exploration.

We do not have hardware to even confirm this, as in our opinion this is supposed to work just fine. However, I open an issue in case somebody sees this and provides a solution. 

See this discussion: https://github.com/acidanthera/AppleALC/pull/485.

CC @Sniki

Please refrain from commenting unless you are developer willing to take this issue.

**Comment by singhalrishi27** [comment]
Yes my idt codec doesn't work with Apple ALC but works oksh with VoodooHDA somehow got alc working by a random dsdt but it is not a permanent solution can you help me to extract patches. I can open an issue but after seeing you hard work I don't wanted to disturb you for some random things

**Comment by Mirone** [comment]
Hi vit, interesting discussion but a pity that was closed, I wanted to be able to follow and give my 2 cents of contribution, of course if it is in accordance with my knowledge and skills,
Best regards!

Mirone

**Comment by singhalrishi27** [comment]
@Mirone 
No, it was not Solved

**Comment by Mirone** [comment]
@singhalrishi27 
I'm pretty rusty for AppleHDA issues, and no hardware for testing, but I want to have a look at that.

**Comment by singhalrishi27** [comment]
@Mirone 
I have IDT92HD91BXX Codec
If you need to test something you can just order me I will do that for you and will provide logs
Acidanthera is working hard if I can help it would be a pleasure to me😄😄😄😍

**Comment by Mirone** [comment]
Of course, you can send me a PM on Insanelymac if you prefer, and we can test something as soon as possible.

**Comment by singhalrishi27** [comment]
@Mirone 
Yup can we go on instagram or telegram if you have no problems

**Comment by Mirone** [comment]
@singhalrishi27 ,
Currently I do not use either, contact me at Olarila.com or Insanelymac.com and I will answer you.

**Comment by singhalrishi27** [comment]
@Mirone 
Done I have DMed you check it out

**Comment by Mirone** [comment]
Thanks!

**Comment by Sniki** [comment]
Appearently CoreBoot has been updated a few days ago and they published the hda-verb extracted from Lenovo ThinkPad T440P OEM Firmware.

Hda-Verb:
https://github.com/coreboot/coreboot/blob/master/src/mainboard/lenovo/t440p/hda_verb.c

GPIO:
https://github.com/coreboot/coreboot/blob/master/src/mainboard/lenovo/t440p/gpio.c

Is there anything that we can extract from this that may fix the issue  ?

CC @wern-apfel @roddy20

**Comment by wern-apfel** [comment]
Some verb overrides. I'm not sure if it helps, but you can try sending the HP mute command (0x0153b080)

Another solution could be [alc-verb](https://github.com/black-dragon74/AppleALC/tree/alc-verb) with a startup script.
The reason why it works with CodecCommander and not when sending the commands via AppleALC could be the delay. CC is loaded later and overrides the values. At least I think so.

I'm not sure vit will be happy if we comment here (see footer)

**Comment by Sniki** [comment]
@black-dragon74

Do you have any recommendation on where we should place alc-verb bin file ?
/usr/local/bin ?

I plan to update ALCPlugFix so it works using alc-verb now instead of CC or VerbStub.

Plan to do a clean install on my T440S with latest changes and see if audio problem has been fully resolved now.

I believe it should be finally solved with this update.

Thanks for your great work.

**Comment by black-dragon74** [comment]
> Do you have any recommendation on where we should place alc-verb bin file ?
> /usr/local/bin ?

Yes. `/usr/local/bin` should be the preferred location.

> I plan to update ALCPlugFix so it works using alc-verb now instead of CC or VerbStub.

You might wanna check: https://github.com/black-dragon74/ALCPlugFix-Swift, it doesn't require `alc-verb` or `hda-verb` dependency and can be configured using a sample PLIST and supports sending verbs on headset plug, unplug, system boot and sleep/wake cycle. For an idea of a sample config file, take a look at: https://github.com/black-dragon74/ALCPlugFix-Swift/blob/master/ALCPlugFix-Swift/Misc/alc_295.plist

Regards

**Comment by Sniki** [comment]
> > Do you have any recommendation on where we should place alc-verb bin file ?
> > /usr/local/bin ?
> 
> Yes. `/usr/local/bin` should be the preferred location.
> 
> > I plan to update ALCPlugFix so it works using alc-verb now instead of CC or VerbStub.
> 
> You might wanna check: https://github.com/black-dragon74/ALCPlugFix-Swift, it doesn't require `alc-verb` or `hda-verb` dependency and can be configured using a sample PLIST and supports sending verbs on headset plug, unplug, system boot and sleep/wake cycle. For an idea of a sample config file, take a look at: https://github.com/black-dragon74/ALCPlugFix-Swift/blob/master/ALCPlugFix-Swift/Misc/alc_295.plist
> 
> Regards

What about alc-verb added into AppleALC, how would we make proper use of it, it would be nice if we add a little sample in readme or a different doc file.

Im sure the entire AppleALC could see a decent amount of codec configuration cleanups (all these patches that are sent as PR in use with CC or ALCPlugFix)

**Comment by black-dragon74** [comment]
> What about alc-verb added into AppleALC, how would we make proper use of it, it would be nice if we add a little sample in readme or a different doc file.

Well, the way `ALCPlugFix-Swift` is designed, installing of `alc-verb` is not needed. However one may use it for debugging purposes in the same way as `hda-verb`. Nothing has changed, you just have to substitute `hda-verb` with `alc-verb`. The new name was chosen to remove the ambiguity between two different user-clients (one requires `CodecCommander` while the other requires `AppleALC`).

> it would be nice if we add a little sample in readme or a different doc file.

There exists a sample configuration file in the **Misc** folder of the repo. It can be viewed [here](https://github.com/black-dragon74/ALCPlugFix-Swift/blob/master/ALCPlugFix-Swift/Misc/sample.plist)

Once you prepare the config file for your codec you just need to run `install.sh` located in the release dist. It will verify the config, bootstrap and install the daemon.

P.S: There exists a way in `ALCPlugFix-Swift` to read the command output however it is not recommended as the commands are sent only on the specified events. To see it in action, just run the binary with your PLIST file as it's first argument. One is better off using `alc-verb` for such purposes.

Regards

**Comment by Sniki** [comment]
Some decent changes and new features have been added to AppleALC but I couldn't find a clear documentation for them.
The one that I am the most interested into is the 'alcverbs=1' or alc-verbs DeviceProperty.

How should we correctly use it and what is it meant to actually do, which custom verbs does it send  ?, etc. does it require alc-verb ?
It would be nice to have a short description of boot-args / DeviceProperties at least in readme.md like other Acidanthera projects have.

Is it supposed to be an alternative or replacement for ALCPlugFIx so it does solve the problem with codecs like ALC3232(ALC292) static noise which right now no real solution is available aside from using ALCPlugFix which this issue is opened for ?

**Comment by wern-apfel** [comment]
> Some decent changes and new features have been added to AppleALC but I couldn't find a clear documentation for them.
> The one that I am the most interested into is the 'alcverbs=1' or alc-verbs DeviceProperty.
> 
> How should we correctly use it and what is it meant to actually do, which custom verbs does it send ?, etc. does it require alc-verb ?
> It would be nice to have a short description of boot-args / DeviceProperties at least in readme.md like other Acidanthera projects have.
> 
> Is it supposed to be an alternative or replacement for ALCPlugFIx so it does solve the problem with codecs like ALC3232(ALC292) static noise which right now no real solution is available aside from using ALCPlugFix which this issue is opened for ?

alc-verb is a replacement for hda-verb and meant to be used without CodecCommander. So, add alcverbs=1 as boot-arg, copy alc-verb to usr/local/bin and use it like hda-verb. "alc-verb 0x0 0xf00 0x0" should show you the vendor/device-id.

**Comment by roddy20** [comment]
>
> 
> How should we correctly use it and what is it meant to actually do, which custom verbs does it send ?, 

all the verbs are described here https://www.intel.com/content/dam/www/public/us/en/documents/product-specifications/high-definition-audio-specification.pdf

**Comment by andreszerocross** [comment]
@wern-apfel i can confirm without CodecCommander or Alc-plugfix audio still work with layout id 4 for Creative CA0132.

5.1 sound still work

**Comment by wern-apfel** [comment]
> @wern-apfel i can confirm without CodecCommander or Alc-plugfix audio still work with layout id 4 for Creative CA0132.
> 
> 5.1 sound still work

Yes, it works since the last commit but I'm not sure if it's fully working after sleep. And the switcher requires the CodecCommander. I already have a switcher based on alc-verb but it's randomly changes the device order.

**Comment by andreszerocross** [comment]
Sorry i can't test it. Because my computer always no display after wake from sleep,,

I can hear audio but no output,,

Common problem for Intel HD 530 Skylake desktop without dGPU

**Comment by Sniki** [comment]
It's been a while since posted this issue and i wanted to give another shot at this problem, i came up with the idea of checking what happens if i do separate codec dumps on linux.
1. codec_dump.txt - the standard codec dump.
2. codec_dump_headphones.txt - a codec dump with headset plugged into the laptop headphone jack.
3. codec_dump_headphones_dock - a codec dump with headset plugged into the docking station headphone jack.

3. can be ignored, uploaded it just in case it would help if anything.

After comparing the 1 & 2 with diffmerge, i noticed that some amp in and amp out values change and strange enough (if that's normal), i noticed the audio selector node 0x22 switches connection asterix symbol from node 0x12 to 0x1a.

Hopefully this gives some additional information to the audio professionals.

[codec_dump.txt](https://github.com/acidanthera/bugtracker/files/10395168/codec_dump.txt)
[codec_dump_headphones.txt](https://github.com/acidanthera/bugtracker/files/10395169/codec_dump_headphones.txt)
[codec_dump_headphones_dock.txt](https://github.com/acidanthera/bugtracker/files/10395170/codec_dump_headphones_dock.txt)
